refactor Client

Deric-W opened the pull request:
Hi,
I took a look at the current `Client` implementation and found a few opportunities for improvement:

| Change                                         | Reason                                                                           |
|------------------------------------------------|----------------------------------------------------------------------------------|
| refactor WaitSet                               | allow waiting for Clients                                                        |
| return a Task containing the Response          | allow for async waiting, prevents deadlocks when both Client and Service run in Unity |
| replace WaitForService with IsServiceAvailable | allow users to handle polling themselves, for example using `WaitUntil` in Unity |
| change default Qos Profile to SERVICES_DEFAULT |  seems to be [defined](https://docs.ros.org/en/rolling/Concepts/About-Quality-of-Service-Settings.html#qos-profiles) specifically for services|
| add tests and documentation                    |                                                                                  |

The changes where only tested on ROS2 Humble on Windows.

You commented on the last PR that you where not having enough time, I could create a separate PR containing your changes (with attributions of course) on `RobotecAI/ros2cs` if you want.

Deric-W committed:
refactor WaitSet to allow waiting for clients

Deric-W committed:
refactor Client.SendAndRecv

 - add clients to WaitSet
 - return a Task to allow for async waiting
 - convert IntPtr to Message

Deric-W committed:
configure C# framework in CMakeLists.txt

Deric-W committed:
handle remaining Tasks at disposal of Client

If a Client gets disposed all of its remaining Tasks
are set to Faulted.

Deric-W committed:
rename Client.SendAndRecv and add sync version

Deric-W committed:
add Client tests

Deric-W committed:
fix Client disposal

Deric-W committed:
add tests for Service

Deric-W committed:
replace Client.WaitForService

The replacement Client.IsServiceAvailable is a simple check
which allows users to handle polling themselves, like using
coroutines in Unity.

Deric-W committed:
add test for Client.IsServiceAvailable

Deric-W committed:
change default Client/Service Qos Profile to SERVICES_DEFAULT

Deric-W committed:
add documentation

Deric-W committed:
fix race condition in Client.CallAsync

Deric-W committed:
replace `default` literals

Occurrences have been replaced with the `default` operator
to prevent build failures with older C# versions.

Deric-W commented:
I noticed a potential resource leak when "abandoning" `Task`s created by `Client.CallAsync`, please don't merge this PR yet.

miyakoshi-dev commented:
Hi.
that's a great improvement
Since the amount is so large that I don't have enough time to check and test everything at once, could you please make a separate PR for your proposal?

Deric-W committed:
Add `Client.PendingRequests` and `Client.Cancel`

These changes are intended to prevent pending Tasks
which are "forgotten" by the user from accumulating
in the Client until they are completed.

Deric-W committed:
make `Client.PendingTasks`thread safe

Deric-W committed:
dispose clients and services on removal

Inspired by #37

Deric-W committed:
change target framework to `netstandard2.0`

This is required for working on all supported Unity versions

Deric-W committed:
fix error when spinning without things to wait on

SpinOnce will return false in this case and Spin
will sleep until the timeout has elapsed.
